var/spack/repos/builtin/packages/singularity/package.py

```diff
@@ -42,23 +42,24 @@ def gopath(self):
 
     @property
     def sylabs_gopath_dir(self):
-        return join_path(self.gopath, 'src/github.com/sylabs/')
+        return join_path(self.stage.source_path, 'github.com/sylabs/')
 
     @property
     def singularity_gopath_dir(self):
         return join_path(self.sylabs_gopath_dir, 'singularity')
 
-    # Unpack the tarball as usual, then move the src dir into
-    # its home within GOPATH.
+    # Unpack the tarball as usual but ensure it lands in its home within
+    # GOPATH.
     def do_stage(self, mirror_only=False):
         super(Singularity, self).do_stage(mirror_only)
-        source_path = self.stage.source_path
         if not os.path.exists(self.singularity_gopath_dir):
+            go_source_path = join_path(self.stage.path, 'go')
+            tty.debug("Temporarily moving {0} to {1}".format(
+                self.stage.source_path, go_source_path))
+            shutil.move(self.stage.source_path, go_source_path)
             tty.debug("Moving {0} to {1}".format(
-                source_path, self.singularity_gopath_dir))
-            mkdirp(self.sylabs_gopath_dir)
-            shutil.move(source_path,
-                        self.singularity_gopath_dir)
+                go_source_path, self.singularity_gopath_dir))
+            shutil.move(go_source_path, self.singularity_gopath_dir)
```

Comment:
Why is there now a double move with shutil?

Reply:
IIRC I did it to ensure the `zip` and `tar.gz` files reside in the same directory as they did before the change.

Reply:
Hopefully this is explained above and/or in the comments I'm pushing.

Reply:
Thanks Tamama! I'll be able to look more closely probably tomorrow, I'm just pooped for today :) I'll also ping @harzell since he's king of the Python debugger (and we worked on the original together)!

Reply:
okay taking a look now! I'm setting the debugger at the beginning of here:

```python
    # Unpack the tarball as usual but ensure it lands in its home within
    # GOPATH.
    def do_stage(self, mirror_only=False):
        super(Singularity, self).do_stage(mirror_only)
        import pdb; pdb.set_trace()
...
```
Then to trigger I'm building `singularity@develop`
```bash
bin/spack install singularity@develop
```
The self.stage.source_path is the "src' folder in the build directory:
```python
(Pdb) self.stage.source_path
'/tmp/spack/var/spack/stage/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv/src'
```
If we look there, we see Singularity extracted
```bash
c$ ls
CHANGELOG.md     COPYRIGHT.md  etc       INSTALL.md       makeit   README.md
cmd              dist          examples  internal         mconfig  scripts
CONTRIBUTING.md  docs          go.mod    LICENSE-LBNL.md  mlocal   vendor
CONTRIBUTORS.md  e2e           go.sum    LICENSE.md       pkg
```
In one folder up, it is empty:

```bash
$ ls ../
src
```
Now let's step forward in the debugger. The next check is to see if the (subfolder) exists that the GOPATH expects Singularity to be in (because we are building relative to where we extracted Singularity, and the repo needs to be on this GOPATH.
```python
(Pdb) self.singularity_gopath_dir
'/tmp/spack/var/spack/stage/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv/src/github.com/sylabs/singularity'
(Pdb) os.path.exists(self.singularity_gopath_dir)
False
```
Note that it falls within the same temporary folder in src, but has an additional name for `github.com/sylabs/singularity`. It doesn't exist, so when we move forward in the debugger we will step into the if statement. The go source path is now being set to the same directory.

```python
/tmp/spack/var/spack/stage/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv/go
```
This is kind of weird, because (if this is referring to the path where go is installed) this would sit here:

```python
os.system('which go')
/tmp/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/go-1.11.5-gmcvqdwr6az4hxbicr375if5fttk6sna/bin/go
```
Then the directory (src) is essentially renamed to go:
```python
$ ls
go
vanessa@vanessa-ThinkPad-T460s:/tmp/spack/var/spack/stage/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv$ ls go/
CHANGELOG.md     COPYRIGHT.md  etc       INSTALL.md       makeit   README.md
cmd              dist          examples  internal         mconfig  scripts
CONTRIBUTING.md  docs          go.mod    LICENSE-LBNL.md  mlocal   vendor
CONTRIBUTORS.md  e2e           go.sum    LICENSE.md       pkg
```
And then we rename it again...

```python
$ ls src/github.com/sylabs/singularity
CHANGELOG.md     COPYRIGHT.md  etc       INSTALL.md       makeit   README.md
cmd              dist          examples  internal         mconfig  scripts
CONTRIBUTING.md  docs          go.mod    LICENSE-LBNL.md  mlocal   vendor
CONTRIBUTORS.md  e2e           go.sum    LICENSE.md       pkg
```
And install proceeds okay.

So - it works, but I don't totally follow why we are doing multiple renames here. It makes sense that we can't rename into a subdirectory, so why are we cloning into src to begin with? With the previous method, we could clone into the stage directory, create the parent of the final folder, and do one clean shutil.move. Why is this no longer an option?

Reply:
I'd like to get @hartzell take on this too.

Reply:
> The self.stage.source_path is the "src' folder in the build directory:
> 
> ```python
> (Pdb) self.stage.source_path
> '/tmp/spack/var/spack/stage/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv/src'
> ```
> 
> If we look there, we see Singularity extracted
> 
> ```shell
> c$ ls
> CHANGELOG.md     COPYRIGHT.md  etc       INSTALL.md       makeit   README.md
> cmd              dist          examples  internal         mconfig  scripts
> CONTRIBUTING.md  docs          go.mod    LICENSE-LBNL.md  mlocal   vendor
> CONTRIBUTORS.md  e2e           go.sum    LICENSE.md       pkg
> ```

The point of the original work,  #11528, was to make `stage.source_path` available before it is created.  This was addressed by using a "well-known" source directory (i.e., appending `src` to the value of `stage.path`.  The source is then extracted/expanded into that directory.


> ```python
> (Pdb) self.singularity_gopath_dir
> '/tmp/spack/var/spack/stage/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv/src/github.com/sylabs/singularity'
> (Pdb) os.path.exists(self.singularity_gopath_dir)
> False
> ```
> 
> Note that it falls within the same temporary folder in src, but has an additional name for `github.com/sylabs/singularity`. It doesn't exist, so when we move forward in the debugger we will step into the if statement. The go source path is now being set to the same directory.
> 
> ```python
> /tmp/spack/var/spack/stage/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv/go
> ```

I was leveraging what I assumed was a well-known `go` path based on `stage.GoFetchStrategy.fetch()`.

> 
> This is kind of weird, because (if this is referring to the path where go is installed) this would sit here:
> 
> ```python
> os.system('which go')
> /tmp/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/go-1.11.5-gmcvqdwr6az4hxbicr375if5fttk6sna/bin/go
> ```

Using the location of the executable as a path would be weird.

> 
> Then the directory (src) is essentially renamed to go:
> 
> ```python
> $ ls
> go
> vanessa@vanessa-ThinkPad-T460s:/tmp/spack/var/spack/stage/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv$ ls go/
> CHANGELOG.md     COPYRIGHT.md  etc       INSTALL.md       makeit   README.md
> cmd              dist          examples  internal         mconfig  scripts
> CONTRIBUTING.md  docs          go.mod    LICENSE-LBNL.md  mlocal   vendor
> CONTRIBUTORS.md  e2e           go.sum    LICENSE.md       pkg
> ```
> 
> And then we rename it again...
> 
> ```python
> $ ls src/github.com/sylabs/singularity
> CHANGELOG.md     COPYRIGHT.md  etc       INSTALL.md       makeit   README.md
> cmd              dist          examples  internal         mconfig  scripts
> CONTRIBUTING.md  docs          go.mod    LICENSE-LBNL.md  mlocal   vendor
> CONTRIBUTORS.md  e2e           go.sum    LICENSE.md       pkg
> ```
> 
> And install proceeds okay.
> 
> So - it works, but I don't totally follow why we are doing multiple renames here. It makes sense that we can't rename into a subdirectory, so why are we cloning into src to begin with? 

The `src` directory is supposed to be the (new) "well-known" destination directory for fetched/expanded source.

> With the previous method, we could clone into the stage directory, create the parent of the final folder, and do one clean shutil.move. Why is this no longer an option?

The problem is the well-known source destination (`stage.source_path`), which is automatically used as the clone destination by the parent `do_stage` method, is now an ancestor of `self.singularity_gopath_dir` so cannot be moved to that location.

Reply:
For the go source path, I was pointing out that the GO_PATH (a subfolder of where it's installed ) is already at `
/tmp/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/go-1.11.5-gmcvqdwr6az4hxbicr375if5fttk6sna/` (just removed bin/go from there, I just used go to show where it was installed)

Reply:
> The src directory is supposed to be the (new) "well-known" destination directory for fetched/expanded source.
So what happens when the build needs to add other content to the go path under src?

Reply:
> For the go source path, I was pointing out that the GO_PATH (a subfolder of where it's installed ) is already at ` /tmp/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/go-1.11.5-gmcvqdwr6az4hxbicr375if5fttk6sna/` (just removed bin/go from there, I just used go to show where it was installed)

Thanks for the clarification.

Reply:
> > The src directory is supposed to be the (new) "well-known" destination directory for fetched/expanded source.
> 
> So what happens when the build needs to add other content to the go path under src?

Excellent question.  Let me confer with Todd.

Reply:
@vsoch -- I'm unsure what you're hung up on here (asked while fully remember the last time I was confused about something with this package :sweat_smile: ...).

Here's what I think is going on:

- in the olden days, the distribution tarball was unpacked into something based on its spack name, e.g. after staging emacs, I have:

    ```
    $ ls
    emacs-26.1  emacs-26.1.tar.gz
    $ ls emacs-26.1
    aclocal.m4  ChangeLog.1   CONTRIBUTE  leim     make-dist    oldXMenu
    admin       ChangeLog.2   COPYING     lib      Makefile.in  README
    autogen.sh  ChangeLog.3   doc         lib-src  modules      site-lisp
    BUGS        config.bat    etc         lisp     msdos        src
    build-aux   configure     info        lwlib    nextstep
    ChangeLog   configure.ac  INSTALL     m4       nt
    $
    ```

- but now, using this PR the source is uniformly unpacked under `src`

    ```
    $ ls
    emacs-26.1.tar.gz  src
    $ ls src
    aclocal.m4  build-aux    ChangeLog.3   CONTRIBUTE  info     lib-src  make-dist    nextstep  site-lisp
    admin       ChangeLog    config.bat    COPYING     INSTALL  lisp     Makefile.in  nt        src
    autogen.sh  ChangeLog.1  configure     doc         leim     lwlib    modules      oldXMenu
    BUGS        ChangeLog.2  configure.ac  etc         lib      m4       msdos        README
    ```
- go is a unique and special child (and currently my favorite language), which has its own gopher-ish way of doing things.  In particularly, it wants to see the the source for something beneath $GOPATH/src, at its import path (`$GOPATH/src/github.com/sylabs/singularity`).  Keep in mind that we fudge $GOPATH to point into `self.stage.path` via the `gopath` property.

- this change to the singularity package:
    -  moves the unpacked tarball dir from `src` (where the default machinery left it for us) to something temporary so that we can set `src` up the way we need it to be;
    - creates the necessary path; and
    - moves the unpacked tarball (`src`) into where it needs to be (`src/github.com/sylabs/singularity`).

If some other bit, e.g. perhaps a resource, needed to put something in a place where `go` would find it, that place would be `$GOPATH/src/<host>/<org>/<package>`, and the code that did so could easily build up that path starting at `self.stage.source_path` (or build it up from the `self.gopath` [aka `self.stage.path`]).

When you ask this:

> So what happens when the build needs to add other content to the go path under src?

is there something you're thinking about that *wouldn't* just go nicely into the tree beneath `src`?

If I were going to change *anything* about this new bit of code, I would name the temporary subdirectory inside `self.stage.path` something other than `go` that made it clear that it was just a temporary directory required for filesystem hoop-jumping, perhaps `hoop-jumping-temp-name` (although that might be a tad too sarcastic...).

Reply:
@hartzell I was first hung up on having a double rename, and I realize it's necessary so the change is warranted. I agree that a different name than "go" would be most appropriate to indicate the directory should quickly go away. A standard [tempfile.mkdtemp](https://docs.python.org/2/library/tempfile.html#tempfile.mkdtemp) with the prefix in the stage directory could work.

The situation I'm thinking about is for go dependencies. Let's say I have one library, The deps will be downloaded into the same staging folder, and cleaned up when we finish.

When it finishes, we don't keep any of those files anyway:

```bash
$ ls /tmp/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/singularity-develop-2m3xjwdj26znt6q5khon4aaeuxiky5zv/
```

Given how go works with a common root, I'm trying to get my head around what it means for spack. I might just be inexperienced (and this is ok) but for other software, it seems like we keep libraries around:

```bash
$ ls /tmp/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/openssl-1.1.1b-ebwq42vxmh5nrsdev4rbk6i4ph4w4fex/include/openssl/
aes.h       buffererr.h  cryptoerr.h  ecdsa.h      kdf.h         opensslconf.h  rand.h       ssl2.h      uierr.h
asn1err.h   buffer.h     crypto.h     ecerr.h      lhash.h       opensslv.h     rc2.h        ssl3.h      ui.h
asn1.h      camellia.h   cterr.h      ec.h         md2.h         ossl_typ.h     rc4.h        sslerr.h    whrlpool.h
asn1_mac.h  cast.h       ct.h         engineerr.h  md4.h         pem2.h         rc5.h        ssl.h       x509err.h
asn1t.h     cmac.h       des.h        engine.h     md5.h         pemerr.h       ripemd.h     stack.h     x509.h
asyncerr.h  cmserr.h     dherr.h      e_os2.h      mdc2.h        pem.h          rsaerr.h     storeerr.h  x509v3err.h
async.h     cms.h        dh.h         err.h        modes.h       pkcs12err.h    rsa.h        store.h     x509v3.h
bioerr.h    comperr.h    dsaerr.h     evperr.h     objectserr.h  pkcs12.h       safestack.h  symhacks.h  x509_vfy.h
bio.h       comp.h       dsa.h        evp.h        objects.h     pkcs7err.h     seed.h       tls1.h
blowfish.h  conf_api.h   dtls1.h      hmac.h       obj_mac.h     pkcs7.h        sha.h        tserr.h
bnerr.h     conferr.h    ebcdic.h     idea.h       ocsperr.h     rand_drbg.h    srp.h        ts.h
bn.h        conf.h       ecdh.h       kdferr.h     ocsp.h        randerr.h      srtp.h       txt_db.h
```
And so if something else needs to be compiled, we still have that for reproducibility. For go, the junk in the GOPATH is (sort of) akin to that. But what we are doing for each install is throwing it away. So my question is - is this the proper way to handle installing go dependencies? For example, here is my github.com folder in my (host) GOPATH:

```bash
$ tree /home/vanessa/Documents/Dropbox/Code/Google/go/src/github.com/ -L 1
/home/vanessa/Documents/Dropbox/Code/Google/go/src/github.com/
├── ajg
├── alecthomas
├── blang
├── blevesearch
├── boltdb
├── BurntSushi
├── cockroachdb
├── containers
├── couchbase
├── cpuguy83
├── docker
├── dustin
├── edsrzf
├── etcd-io
├── fatih
├── fsnotify
├── ghodss
├── glycerine
├── gobuffalo
├── gofrs
├── gogs
├── golang
├── GoogleContainerTools
├── gophercises
├── gorilla
├── go-sql-driver
├── hashicorp
├── jackc
├── jmoiron
├── joho
├── karrick
├── kballard
├── lib
├── magiconair
├── markbates
├── Masterminds
├── mattn
├── microcosm-cc
├── mitchellh
├── monoculum
├── mtrmac
├── nicksnyder
├── nilslice
├── opencontainers
├── pelletier
├── philhofer
├── pkg
├── rancher
├── RoaringBitmap
├── rogpeppe
├── sci-f
├── serenize
├── sergi
├── sirupsen
├── sourcegraph
├── spf13
├── steveyen
├── stretchr
├── sylabs
├── tidwall
├── tinylib
├── unrolled
├── urfave
├── vbatts
└── vsoch
```
A lot of libraries use urfave for the client, for example, so with our setup we are downloading and throwing away these each time. Maybe this is best, given that we get some "fresh" version, and each install is modular. But it seems to not mirror how the user is expected to maintain repos on the local machine.

All of this is changed with [modules](https://blog.golang.org/using-go-modules) because we can have an install that isn't located inside the GOPATH. What we should discuss (and possibly out of scope for this PR) is if spack should generate new GOPATH each time, or have some common or shared GOPATH to maintain downloads of the software (more reproducible if we need them again?)

Hopefully I've better articulated myself here :)

Reply:
@hartzell I see your other comment via email (and I'm lost in a forest here and can't find it!) but this, this, yes!

> The proper Spack thing to do is probably going to be to create "resources" for each of the things listed in the go.mod file (and recursively) and write "resource stanzas" in the package.py file. Then spack fetch [SIC?] can go and get them, tuck them into the appropriate dir, and away things will go.
> caddy has 16 things listed in its go.mod but when you go mod download them, you end up with 46 things (recursive...). Other tools have even bigger combinatorial explosions....
> Unlike other languages that do this (e.g. lua-rocks), Go goes to great lengths to ensure that the thing it downloads is the thing that it expects. We'll have a file full of checksums (go.sum) within the tarball that we checksum ourselves.
> Perhaps instead of writing resource definitions, Spack can leverage the existing Go machinery somehow?
> I think that Spack-folk will need to kick this around a bit....

Go is feeling a *lot* like NPM with dependencies, and it's a mess in terms of trying to fit that into spack space, where everything is understood, versioned, and check summed. This, on the other hand, is like a free for all. I think this is probably out of scope for this PR (we are just doing a tiny update for an already existing strategy to install, and it works) but +1 this is something the spack gods / developers need to talk more about!

Reply:
@tldahlgren -- seems like @vsoch's comfortable where this has ended up (at the risk of putting words in her mouth).  Don't let the can of worms that I popped the top off of above keep you from marking this conversation resolved.  I'd do it myself, but since I didn't start it....

Reply:
I'm good to resolve it, but @hartzell let's remember where this discussion is if we need it!